src/modules/ekf2/EKF/mag_3d_control.cpp

```diff
@@ -0,0 +1,215 @@
+		ECL_INFO("starting mag 3D fusion");
+
+	} else if (!_control_status.flags.mag_3D && _control_status_prev.flags.mag_3D) {
+		ECL_INFO("stopping mag 3D fusion");
+	}
+
+	// if we are using 3-axis magnetometer fusion, but without external NE aiding,
+	// then the declination must be fused as an observation to prevent long term heading drift
+	// fusing declination when gps aiding is available is optional, but recommended to prevent
+	// problem if the vehicle is static for extended periods of time
+	const bool mag_decl_user_selected = (_params.mag_declination_source & GeoDeclinationMask::FUSE_DECL);
+	const bool not_using_ne_aiding = !_control_status.flags.gps;
+	_control_status.flags.mag_dec = (_control_status.flags.mag && (not_using_ne_aiding || mag_decl_user_selected));
+
+	if (_control_status.flags.mag) {
+		aid_src.timestamp_sample = mag_sample.time_us;
+		aid_src.fusion_enabled = true;
+
+		if (continuing_conditions_passing && _control_status.flags.yaw_align) {
+
+			if (mag_sample.reset || checkHaglYawResetReq() || wmm_updated) {
+				ECL_INFO("reset to %s", AID_SRC_NAME);
+				resetMagStates(_mag_lpf.getState(), _control_status.flags.mag_hdg || _control_status.flags.mag_3D);
+				aid_src.time_last_fuse = _time_delayed_us;
+
+			} else {
+				if (!_mag_decl_cov_reset) {
+					// After any magnetic field covariance reset event the earth field state
+					// covariances need to be corrected to incorporate knowledge of the declination
+					// before fusing magnetometer data to prevent rapid rotation of the earth field
+					// states for the first few observations.
+					fuseDeclination(0.02f);
+					_mag_decl_cov_reset = true;
+					fuseMag(mag_sample.mag, aid_src, false);
+
+				} else {
+					// The normal sequence is to fuse the magnetometer data first before fusing
+					// declination angle at a higher uncertainty to allow some learning of
+					// declination angle over time.
+					const bool update_all_states = _control_status.flags.mag_3D;
+					fuseMag(mag_sample.mag, aid_src, update_all_states);
+
+					if (_control_status.flags.mag_dec) {
+						fuseDeclination(0.5f);
+					}
+				}
+			}
+
+			const bool is_fusion_failing = isTimedOut(aid_src.time_last_fuse, _params.reset_timeout_max);
+
+			if (is_fusion_failing) {
+				if (_nb_mag_3d_reset_available > 0) {
+					// Data seems good, attempt a reset (mag states only unless mag_3D currently active)
+					ECL_WARN("%s fusion failing, resetting", AID_SRC_NAME);
+					resetMagStates(_mag_lpf.getState(), _control_status.flags.mag_hdg || _control_status.flags.mag_3D);
+					aid_src.time_last_fuse = _time_delayed_us;
+
+					if (_control_status.flags.in_air) {
+						_nb_mag_3d_reset_available--;
+					}
+
+				} else if (starting_conditions_passing) {
+					// Data seems good, but previous reset did not fix the issue
+					// something else must be wrong, declare the sensor faulty and stop the fusion
+					//_control_status.flags.mag_fault = true;
```

Comment:
yes, we could make it faulty here and add a logic to clear the fault if we see that he innovation is continuously low